tasks.go

```diff
@@ -1,125 +1,121 @@
 type Task func() error
 
-// GetAllTasks returns a list of all tasks performed by the dump tool. It may
-// return tasks even in the presence of an error.
-// FIXME: GetAllTasks should not need to know about basepath.
-func GetAllTasks(basepath string) ([]Task, error) {
-	var (
-		tasks     []Task
-		retErrors errorList
-	)
-
-	var (
-		resources = []string{"deploymentconfigs", "pods", "services", "events"}
-		// We should only care about logs for pods, because they cover
-		// all other possible types.
-		resourcesWithLogs = []string{"pods"}
-	)
-
-	projects, err := GetProjects()
-	if err != nil {
-		return nil, err
-	}
-
-	if len(projects) == 0 {
-		return nil, errors.New("no projects visible to the currently logged in user")
-	}
+// NewError returns a Task that always return the given error.
+func NewError(err error) Task {
+	return func() error { return err }
+}
 
-	// Add tasks to fetch resource definitions.
-	definitionsTasks, err := GetResourceDefinitionsTasks(projects, resources, basepath)
-	if err != nil {
-		retErrors = append(retErrors, err)
-	}
-	tasks = append(tasks, definitionsTasks...)
+var (
+	resources = []string{"deploymentconfigs", "pods", "services", "events"}
+	// We should only care about logs for pods, because they cover all other
+	// possible types.
+	resourcesWithLogs = []string{"pods"}
+)
 
-	// Add tasks to fetch logs.
-	logsTasks, err := GetFetchLogsTasks(projects, resourcesWithLogs, basepath)
-	if err != nil {
-		retErrors = append(retErrors, err)
-	}
-	tasks = append(tasks, logsTasks...)
+// GetAllTasks returns a list of all tasks known to the dump tool.
+// FIXME: GetAllTasks should not need to know about basepath.
+func GetAllTasks(basepath string) <-chan Task {
+	tasks := make(chan Task)
+	go func() {
+		defer close(tasks)
 
-	nagiosDataTasks, err := GetNagiosTasks(projects, basepath)
-	if err != nil {
-		retErrors = append(retErrors, err)
-	}
-	tasks = append(tasks, nagiosDataTasks...)
+		projects, err := GetProjects()
+		if err != nil {
+			tasks <- NewError(err)
+			return
+		}
+		if len(projects) == 0 {
+			tasks <- NewError(errors.New("no projects visible to the currently logged in user"))
+			return
+		}
 
-	// Add check tasks
-	for _, p := range projects {
-		outFor := outToFile(basepath, "json", "analysis")
-		errOutFor := outToFile(basepath, "stderr", "analysis")
-		task := CheckTasks(p, outFor, errOutFor)
-		tasks = append(tasks, task)
-	}
+		var wg sync.WaitGroup
+
+		// Add tasks to fetch resource definitions.
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			GetResourceDefinitionsTasks(projects, resources, basepath, tasks)
+		}()
+
+		// Add tasks to fetch logs.
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			GetFetchLogsTasks(projects, resourcesWithLogs, basepath, tasks)
+		}()
+
+		// Add tasks to fetch Nagios data.
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			GetNagiosTasks(projects, basepath, tasks)
+		}()
+
+		// Add check tasks.
+		wg.Add(1)
+		go func() {
```

Comment:
Good question. We can definitely arrange for them to run after the others.

Do you prefer we do it now, or just add a note in case that condition becomes true?